README.md
# simpleeval (Simple Eval)

[![Build Status](https://travis-ci.org/danthedeckie/simpleeval.svg?branch=master)](https://travis-ci.org/danthedeckie/simpleeval)  [![Coverage Status](https://coveralls.io/repos/danthedeckie/simpleeval/badge.png?branch=master)](https://coveralls.io/r/danthedeckie/simpleeval?branch=master)

A quick single-file MIT-Licenced library for easily adding evaluatable expressions into python
projects.  Say you want to allow a user to set an alarm volume, which could depend
on the time of day, alarm level, how many previous alarms had gone off, and if there
is music playing at the time.

Or if you want to allow simple formulae in a web application, but don't want to
give full `eval()` access, or don't want to run in javascript on the client side.

It's deliberately very simple, just a single file you can dump into a project, or import
from pypi (pip or easy_install).

Internally, it's using the amazing python `ast` module to parse the expression, which
allows very fine control of what is and isn't allowed.  It should be completely safe in terms
of what operations can be performed by the expression.

The only issue I know to be aware of is that you can create an expression which
takes a long time to evaluate, or which evaluating requires an awful lot of memory,
which leaves the potential for DOS attacks.  There is basic protection against this,
and you can lock it down further if you desire. (see the `Operators` section below)

You should be aware of this when deploying in a public setting.

The defaults are pretty locked down and basic, and it's very easy to add whatever
extra specific functionality you need (your own functions, variable/name lookup, etc).

## To install:

Either just copy the simpleeval.py file into your project, or use pip:

```
    pip install simpleeval
```

(or add simpleeval into your requirements.txt, if that's how your project works...)

## Basic Usage

To get very simple evaluating:

```python
    from simpleeval import simple_eval

    simple_eval("21 + 21")
```

returns `42`.

Expressions can be as complex and convoluted as you want:

```python
    simple_eval("21 + 19 / 7 + (8 % 3) ** 9")
```

returns `535.714285714`.

You can add your own functions in as well.

```python
    simple_eval("square(11)", functions={"square": lambda x: x*x})
```

returns `121`.

For more details of working with functions, read [The docs on pypi](https://pypi.python.org/pypi/simpleeval)

### Note:
all further examples use `>>>` to designate python code, as if you are using the python interactive
prompt.

## Limited Power

Also note, the `**` operator has been locked down by default to have a maximum input value
of `4000000`, which makes it somewhat harder to make expressions which go on for ever.  You
can change this limit by changing the `simpleeval.POWER_MAX` module level value to whatever
is an appropriate value for you (and the hardware that you're running on) or if you want to
completely remove all limitations, you can set the `s.operators[ast.Pow] = operator.pow` or make
your own function.

On my computer, `9**9**5` evaluates almost instantly, but `9**9**6` takes over 30 seconds.
Since `9**7` is `4782969`, and so over the `POWER_MAX` limit, it throws a
``NumberTooHigh` exception for you. (Otherwise it would go on for hours, or until the computer
runs out of memory)

## String Safety

There are also limits on string length (100000 characters, `MAX_STRING_LENGTH`).
This can be changed if you wish.

## If Expressions

You can use python style `if x then y else z` type expressions:

```python
    >>> simple_eval("'equal' if x == y else 'not equal'",
                    names={"x": 1, "y": 2})
    'not equal'
```
which, of course, can be nested:

```python
    >>> simple_eval("'a' if 1 == 2 else 'b' if 2 == 3 else 'c'")
    'c'
```

## Functions

You can define functions which you'd like the expressions to have access to:

```python
    >>> simple_eval("double(21)", functions={"double": lambda x:x*2})
    42
```

You can define "real" functions to pass in rather than lambdas, of course too, and even re-name them so that expressions can be shorter

```python
    >>> def double(x):
            return x * 2
    >>> simple_eval("d(100) + double(1)", functions={"d": double, "double":double})
    202
```

## Names

Sometimes it's useful to have variables available, which in python terminology are called 'names'.

```python
    >>> simple_eval("a + b", names={"a": 11, "b": 100})
    111
```

You can also hand the handling of names over to a function, if you prefer:

```python
    >>> def name_handler(node):
            return ord(node.id[0].lower(a))-96

    >>> simple_eval('a + b', names=name_handler)
    3
```

That was a bit of a silly example, but you could use this for pulling values from a database or file, say, or doing some kind of caching system.

## Creating an Evaluator Class

Rather than creating a new evaluator each time, if you are doing a lot of evaluations,
you can create a SimpleEval object, and pass it expressions each time (which should be a bit quicker, and certainly more convenient for some use cases):

```python
    s = SimpleEval()
    s.eval("1 + 1")
    # and so on...
```
You can assign / edit the various options of the `SimpleEval` object if you want to.
Either assign them during creation (like the `simple_eval` function)

```python
    s = SimpleEval(functions={"boo": boo})
```

or edit them after creation:

```python
    s.names['fortytwo'] = 42
```

this actually means you can modify names (or functions) with functions, if you really feel so inclined:

```python
    s = SimpleEval()
    def set_val(name, value):
        s.names[name.value] = value.value
        return value.value

    s.functions = {'set': set_val}

    s.eval("set('age', 111)")
```
Say.  This would allow a certain level of 'scriptyness' if you had these evaluations happening as callbacks in a program.  Although you really are reaching the end of what this library is intended for at this stage.

## Other...

The library supports both python 2 and 3 using the 2to3 converter.

Please read the `test_simpleeval.py` file for other potential gotchas or details.  I'm very happy to accept pull requests, suggestions, or other issues.  Enjoy!



CHANGELOG.rst
CHANGELOG
#########


Unreleased changes
==================

*   Drop CI testing for Python 2.7 and Python 3.6,
    which have both reached end of life status.


executors.py
import ast

from simpleeval import CompoundEvalWithAssignments, EvalWithAssignments, MAX_COMPREHENSION_LENGTH, IterableTooLong, \
    InvalidExpression

MAX_NUM_STATEMENTS = 100000


class TooManyStatements(InvalidExpression):
    """I've only got so much time to spare!"""
    pass


class SimpleExecutor(EvalWithAssignments):
    def __init__(self, operators=None, functions=None, names=None):
        super(SimpleExecutor, self).__init__(operators, functions, names)
        self._num_stmts = 0

    def eval(self, expr):
        self._num_stmts = 0
        return super(SimpleExecutor, self).eval(expr)

    def exec(self, expr):
        self._num_stmts = 0
        self.expr = expr
        body = ast.parse(expr.strip()).body
        return self._exec(body)

    def _eval(self, node):
        self._num_stmts += 1
        if self._num_stmts > MAX_NUM_STATEMENTS:
            raise TooManyStatements("You are trying to execute too many statements.")

        return super(SimpleExecutor, self)._eval(node)

    def _exec(self, body):
        for expression in body:
            try:
                self._eval(expression)
            except _Return as r:
                return r.value


class ExecutorWithControl(SimpleExecutor, CompoundEvalWithAssignments):
    def __init__(self, operators=None, functions=None, names=None):
        super(ExecutorWithControl, self).__init__(operators, functions, names)

        self.nodes.update({
            ast.If: self._exec_if,
            ast.For: self._exec_for,
            ast.While: self._exec_while,
            ast.Break: self._exec_break,
            ast.Continue: self._exec_continue,
            ast.Return: self._exec_return
        })

    def exec(self, expr):
        self._max_count = 0
        return super(ExecutorWithControl, self).exec(expr)

    def _exec_if(self, node):
        test = self._eval(node.test)
        if test:
            self._exec(node.body)
        else:
            self._exec(node.orelse)

    def _exec_for(self, node):
        for item in self._eval(node.iter):
            self._max_count += 1
            if self._max_count > MAX_COMPREHENSION_LENGTH:
                raise IterableTooLong('Too many loops (in for block)')

            self._assign(node.target, self._FinalValue(value=item))
            try:
                self._exec(node.body)
            except _Break:
                break
            except _Continue:
                continue
        else:
            self._exec(node.orelse)

    def _exec_while(self, node):
        while self._eval(node.test):
            self._max_count += 1
            if self._max_count > MAX_COMPREHENSION_LENGTH:
                raise IterableTooLong('Too many loops (in while block)')

            try:
                self._exec(node.body)
            except _Break:
                break
            except _Continue:
                continue
        else:
            self._exec(node.orelse)

    def _exec_break(self, node):
        raise _Break

    def _exec_continue(self, node):
        raise _Continue

    def _exec_return(self, node):
        raise _Return(self._eval(node.value))


# exceptions used to propogate loop-breaking signals
class _Return(BaseException):
    def __init__(self, retval):
        self.value = retval


class _Break(BaseException):
    pass


class _Continue(BaseException):
    pass


simpleeval.py
"""
SimpleEval - (C) 2013-2022 Daniel Fairhead
-------------------------------------

An short, easy to use, safe and reasonably extensible expression evaluator.
Designed for things like in a website where you want to allow the user to
generate a string, or a number from some other input, without allowing full
eval() or other unsafe or needlessly complex linguistics.

-------------------------------------

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

-------------------------------------

Initial idea copied from J.F. Sebastian on Stack Overflow
( http://stackoverflow.com/a/9558001/1973500 ) with
modifications and many improvements.

-------------------------------------
Contributors:
- corro (Robin Baumgartner) (py3k)
- dratchkov (David R) (nested dicts)
- marky1991 (Mark Young) (slicing)
- T045T (Nils Berg) (!=, py3kstr, obj.
- perkinslr (Logan Perkins) (.__globals__ or .func_ breakouts)
- impala2 (Kirill Stepanov) (massive _eval refactor)
- gk (ugik) (Other iterables than str can DOS too, and can be made)
- daveisfera (Dave Johansen) 'not' Boolean op, Pycharm, pep8, various other fixes
- xaled (Khalid Grandi) method chaining correctly, double-eval bugfix.
- EdwardBetts (Edward Betts) spelling correction.
- charlax (Charles-Axel Dein charlax) Makefile and cleanups
- mommothazaz123 (Andrew Zhu) f"string" support, Python 3.8 support
- lubieowoce (Uryga) various potential vulnerabilities
- JCavallo (Jean Cavallo) names dict shouldn't be modified
- Birne94 (Daniel Birnstiel) for fixing leaking generators.
- patricksurry (Patrick Surry) or should return last value, even if falsy.
- shughes-uk (Samantha Hughes) python w/o 'site' should not fail to import.
- KOLANICH packaging / deployment / setup help & << + >> ops
- graingert (Thomas Grainger) packaging / deployment / setup help
- bozokopic (Bozo Kopic) Memory leak fix

-------------------------------------
Basic Usage:

>>> s = SimpleEval()
>>> s.eval("20 + 30")
50

You can add your own functions easily too:

if file.txt contents is "11"

>>> def get_file():
...     with open("file.txt", 'r') as f:
...         return f.read()

>>> s.functions["get_file"] = get_file
>>> s.eval("int(get_file()) + 31")
42

For more information, see the full package documentation on pypi, or the github
repo.

-----------

If you don't need to re-use the evaluator (with it's names, functions, etc),
then you can use the simple_eval() function:

>>> simple_eval("21 + 19")
40

You can pass names, operators and functions to the simple_eval function as
well:

>>> simple_eval("40 + two", names={"two": 2})
42

"""

import ast
import operator as op
import sys
import warnings
from random import random

PYTHON3 = sys.version_info[0] == 3

########################################
# Module wide 'globals'

MAX_STRING_LENGTH = 100000
MAX_COMPREHENSION_LENGTH = 10000
MAX_POWER = 4000000  # highest exponent
MAX_SHIFT = 10000  # highest << or >> (lshift / rshift)
MAX_SHIFT_BASE = int(sys.float_info.max)  # highest on left side of << or >>
DISALLOW_PREFIXES = ["_", "func_"]
DISALLOW_METHODS = ["format", "format_map", "mro"]

# Disallow functions:
# This, strictly speaking, is not necessary.  These /should/ never be accessable anyway,
# if DISALLOW_PREFIXES and DISALLOW_METHODS are all right.  This is here to try and help
# people not be stupid.  Allowing these functions opens up all sorts of holes - if any of
# their functionality is required, then please wrap them up in a safe container.  And think
# very hard about it first.  And don't say I didn't warn you.
# builtins is a dict in python >3.6 but a module before
DISALLOW_FUNCTIONS = {type, isinstance, eval, getattr, setattr, repr, compile, open}
if hasattr(__builtins__, "help") or (
    hasattr(__builtins__, "__contains__") and "help" in __builtins__  # type: ignore
):
    # PyInstaller environment doesn't include this module.
    DISALLOW_FUNCTIONS.add(help)


if PYTHON3:
    # exec is not a function in Python2...
    exec("DISALLOW_FUNCTIONS.add(exec)")  # pylint: disable=exec-used


########################################
# Exceptions:


class InvalidExpression(Exception):
    """Generic Exception"""

    pass


class FunctionNotDefined(InvalidExpression):
    """sorry! That function isn't defined!"""

    def __init__(self, func_name, expression):
        self.message = "Function '{0}' not defined," " for expression '{1}'.".format(
            func_name, expression
        )
        setattr(self, "func_name", func_name)  # bypass 2to3 confusion.
        self.expression = expression

        super(InvalidExpression, self).__init__(self.message)


class NameNotDefined(InvalidExpression):
    """a name isn't defined."""

    def __init__(self, name, expression):
        self.name = name
        self.message = "'{0}' is not defined for expression '{1}'".format(name, expression)
        self.expression = expression

        super(InvalidExpression, self).__init__(self.message)


class AttributeDoesNotExist(InvalidExpression):
    """attribute does not exist"""

    def __init__(self, attr, expression):
        self.message = "Attribute '{0}' does not exist in expression '{1}'".format(
            attr, expression
        )
        self.attr = attr
        self.expression = expression

        super(InvalidExpression, self).__init__(self.message)


class FeatureNotAvailable(InvalidExpression):
    """What you're trying to do is not allowed."""

    pass


class NumberTooHigh(InvalidExpression):
    """Sorry! That number is too high. I don't want to spend the
    next 10 years evaluating this expression!"""

    pass


class IterableTooLong(InvalidExpression):
    """That iterable is **way** too long, baby."""

    pass


class AssignmentAttempted(UserWarning):
    """Assignment not allowed in SimpleEval"""

    pass


class MultipleExpressions(UserWarning):
    """Only the first expression parsed will be used"""

    pass


########################################
# Default simple functions to include:


def random_int(top):
    """return a random int below <top>"""

    return int(random() * top)


def safe_power(a, b):  # pylint: disable=invalid-name
    """a limited exponent/to-the-power-of function, for safety reasons"""

    if abs(a) > MAX_POWER or abs(b) > MAX_POWER:
        raise NumberTooHigh("Sorry! I don't want to evaluate {0} ** {1}".format(a, b))
    return a ** b


def safe_mult(a, b):  # pylint: disable=invalid-name
    """limit the number of times an iterable can be repeated..."""

    if hasattr(a, "__len__") and b * len(a) > MAX_STRING_LENGTH:
        raise IterableTooLong("Sorry, I will not evalute something that long.")
    if hasattr(b, "__len__") and a * len(b) > MAX_STRING_LENGTH:
        raise IterableTooLong("Sorry, I will not evalute something that long.")

    return a * b


def safe_add(a, b):  # pylint: disable=invalid-name
    """iterable length limit again"""

    if hasattr(a, "__len__") and hasattr(b, "__len__"):
        if len(a) + len(b) > MAX_STRING_LENGTH:
            raise IterableTooLong(
                "Sorry, adding those two together would" " make something too long."
            )
    return a + b


def safe_rshift(a, b):  # pylint: disable=invalid-name
    """rshift, but with input limits"""
    if abs(b) > MAX_SHIFT or abs(a) > MAX_SHIFT_BASE:
        raise NumberTooHigh("Sorry! I don't want to evaluate {0} >> {1}".format(a, b))
    return a >> b


def safe_lshift(a, b):  # pylint: disable=invalid-name
    """lshift, but with input limits"""
    if abs(b) > MAX_SHIFT or abs(a) > MAX_SHIFT_BASE:
        raise NumberTooHigh("Sorry! I don't want to evaluate {0} << {1}".format(a, b))
    return a << b


########################################
# Defaults for the evaluator:

DEFAULT_OPERATORS = {
    ast.Add: safe_add,
    ast.Sub: op.sub,
    ast.Mult: safe_mult,
    ast.Div: op.truediv,
    ast.FloorDiv: op.floordiv,
    ast.RShift: safe_rshift,
    ast.LShift: safe_lshift,
    ast.Pow: safe_power,
    ast.Mod: op.mod,
    ast.Eq: op.eq,
    ast.NotEq: op.ne,
    ast.Gt: op.gt,
    ast.Lt: op.lt,
    ast.GtE: op.ge,
    ast.LtE: op.le,
    ast.Not: op.not_,
    ast.USub: op.neg,
    ast.UAdd: op.pos,
    ast.In: lambda x, y: op.contains(y, x),
    ast.NotIn: lambda x, y: not op.contains(y, x),
    ast.Is: lambda x, y: x is y,
    ast.IsNot: lambda x, y: x is not y,
}

DEFAULT_FUNCTIONS = {
    "rand": random,
    "randint": random_int,
    "int": int,
    "float": float,
    # pylint: disable=undefined-variable
    "str": str if PYTHON3 else unicode,  # type: ignore
}

DEFAULT_NAMES = {"True": True, "False": False, "None": None}

ATTR_INDEX_FALLBACK = True


########################################
# And the actual evaluator:


class SimpleEval(object):  # pylint: disable=too-few-public-methods
    """A very simple expression parser.
    >>> s = SimpleEval()
    >>> s.eval("20 + 30 - ( 10 * 5)")
    0
    """

    expr = ""

    def __init__(self, operators=None, functions=None, names=None):
        """
        Create the evaluator instance.  Set up valid operators (+,-, etc)
        functions (add, random, get_val, whatever) and names."""

        if not operators:
            operators = DEFAULT_OPERATORS.copy()
        if not functions:
            functions = DEFAULT_FUNCTIONS.copy()
        if not names:
            names = DEFAULT_NAMES.copy()

        self.operators = operators
        self.functions = functions
        self.names = names

        self.nodes = {
            ast.Expr: self._eval_expr,
            ast.Assign: self._eval_assign,
            ast.AugAssign: self._eval_aug_assign,
            ast.Import: self._eval_import,
            ast.Num: self._eval_num,
            ast.Str: self._eval_str,
            ast.Name: self._eval_name,
            ast.UnaryOp: self._eval_unaryop,
            ast.BinOp: self._eval_binop,
            ast.BoolOp: self._eval_boolop,
            ast.Compare: self._eval_compare,
            ast.IfExp: self._eval_ifexp,
            ast.Call: self._eval_call,
            ast.keyword: self._eval_keyword,
            ast.Subscript: self._eval_subscript,
            ast.Attribute: self._eval_attribute,
            ast.Index: self._eval_index,
            ast.Slice: self._eval_slice,
        }

        # py3k stuff:
        if hasattr(ast, "NameConstant"):
            self.nodes[ast.NameConstant] = self._eval_constant

        # py3.6, f-strings
        if hasattr(ast, "JoinedStr"):
            self.nodes[ast.JoinedStr] = self._eval_joinedstr  # f-string
            self.nodes[
                ast.FormattedValue
            ] = self._eval_formattedvalue  # formatted value in f-string

        # py3.8 uses ast.Constant instead of ast.Num, ast.Str, ast.NameConstant
        if hasattr(ast, "Constant"):
            self.nodes[ast.Constant] = self._eval_constant

        # Defaults:

        self.ATTR_INDEX_FALLBACK = ATTR_INDEX_FALLBACK

        # Check for forbidden functions:

        for f in self.functions.values():
            if f in DISALLOW_FUNCTIONS:
                raise FeatureNotAvailable("This function {} is a really bad idea.".format(f))

    def __del__(self):
        self.nodes = None

    @staticmethod
    def parse(expr):
        """parse an expression into a node tree"""

        parsed = ast.parse(expr.strip())

        body_len = len(parsed.body)
        if body_len > 0:
            if body_len > 1:
                warnings.warn(
                    "'{}' contains multiple expressions. Only the first will be used.".format(
                        expr
                    ),
                    MultipleExpressions,
                )
            return parsed.body[0]
        raise InvalidExpression("Sorry, cannot evaluate empty string")

    def eval(self, expr, previously_parsed=None):
        """evaluate an expresssion, using the operators, functions and
        names previously set up."""

        # set a copy of the expression aside, so we can give nice errors...
        self.expr = expr

        return self._eval(previously_parsed or self.parse(expr))

    def _eval(self, node):
        """The internal evaluator used on each node in the parsed tree."""

        try:
            handler = self.nodes[type(node)]
        except KeyError:
            raise FeatureNotAvailable(
                "Sorry, {0} is not available in this " "evaluator".format(type(node).__name__)
            )

        return handler(node)

    def _eval_expr(self, node):
        return self._eval(node.value)

    def _eval_assign(self, node):
        warnings.warn(
            "Assignment ({}) attempted, but this is ignored".format(self.expr), AssignmentAttempted
        )
        return self._eval(node.value)

    def _eval_aug_assign(self, node):
        warnings.warn(
            "Assignment ({}) attempted, but this is ignored".format(self.expr), AssignmentAttempted
        )
        return self._eval(node.value)

    @staticmethod
    def _eval_import(node):
        raise FeatureNotAvailable("Sorry, 'import' is not allowed.")

    @staticmethod
    def _eval_num(node):
        return node.n

    @staticmethod
    def _eval_str(node):
        if len(node.s) > MAX_STRING_LENGTH:
            raise IterableTooLong(
                "String Literal in statement is too long!"
                " ({0}, when {1} is max)".format(len(node.s), MAX_STRING_LENGTH)
            )
        return node.s

    @staticmethod
    def _eval_constant(node):
        if hasattr(node.value, "__len__") and len(node.value) > MAX_STRING_LENGTH:
            raise IterableTooLong(
                "Literal in statement is too long!"
                " ({0}, when {1} is max)".format(len(node.value), MAX_STRING_LENGTH)
            )
        return node.value

    def _eval_unaryop(self, node):
        return self.operators[type(node.op)](self._eval(node.operand))

    def _eval_binop(self, node):
        return self.operators[type(node.op)](self._eval(node.left), self._eval(node.right))

    def _eval_boolop(self, node):
        to_return = False
        if isinstance(node.op, ast.And):
            for value in node.values:
                to_return = self._eval(value)
                if not to_return:
                    break
        elif isinstance(node.op, ast.Or):
            for value in node.values:
                to_return = self._eval(value)
                if to_return:
                    break
        return to_return

    def _eval_compare(self, node):
        right = self._eval(node.left)
        to_return = True
        for operation, comp in zip(node.ops, node.comparators):
            if not to_return:
                break
            left = right
            right = self._eval(comp)
            to_return = self.operators[type(operation)](left, right)
        return to_return

    def _eval_ifexp(self, node):
        return self._eval(node.body) if self._eval(node.test) else self._eval(node.orelse)

    def _eval_call(self, node):
        if isinstance(node.func, ast.Attribute):
            func = self._eval(node.func)
        else:
            try:
                func = self.functions[node.func.id]
            except KeyError:
                raise FunctionNotDefined(node.func.id, self.expr)
            except AttributeError:
                raise FeatureNotAvailable("Lambda Functions not implemented")

            if func in DISALLOW_FUNCTIONS:
                raise FeatureNotAvailable("This function is forbidden")

        return func(
            *(self._eval(a) for a in node.args), **dict(self._eval(k) for k in node.keywords)
        )

    def _eval_keyword(self, node):
        return node.arg, self._eval(node.value)

    def _eval_name(self, node):
        try:
            # This happens at least for slicing
            # This is a safe thing to do because it is impossible
            # that there is a true exression assigning to none
            # (the compiler rejects it, so you can't even
            # pass that to ast.parse)
            if hasattr(self.names, "__getitem__"):
                return self.names[node.id]
            if callable(self.names):
                return self.names(node)
            raise InvalidExpression(
                'Trying to use name (variable) "{0}"'
                ' when no "names" defined for'
                " evaluator".format(node.id)
            )

        except KeyError:
            if node.id in self.functions:
                return self.functions[node.id]

            raise NameNotDefined(node.id, self.expr)

    def _eval_subscript(self, node):
        container = self._eval(node.value)
        key = self._eval(node.slice)
        # Currently if there's a KeyError, that gets raised straight up.
        # TODO: Should that be wrapped in an InvalidExpression?
        return container[key]

    def _eval_attribute(self, node):
        for prefix in DISALLOW_PREFIXES:
            if node.attr.startswith(prefix):
                raise FeatureNotAvailable(
                    "Sorry, access to __attributes "
                    " or func_ attributes is not available. "
                    "({0})".format(node.attr)
                )
        if node.attr in DISALLOW_METHODS:
            raise FeatureNotAvailable(
                "Sorry, this method is not available. " "({0})".format(node.attr)
            )
        # eval node
        node_evaluated = self._eval(node.value)

        # Maybe the base object is an actual object, not just a dict
        try:
            return getattr(node_evaluated, node.attr)
        except (AttributeError, TypeError):
            pass

        # TODO: is this a good idea?  Try and look for [x] if .x doesn't work?
        if self.ATTR_INDEX_FALLBACK:
            try:
                return node_evaluated[node.attr]
            except (KeyError, TypeError):
                pass

        # If it is neither, raise an exception
        raise AttributeDoesNotExist(node.attr, self.expr)

    def _eval_index(self, node):
        return self._eval(node.value)

    def _eval_slice(self, node):
        lower = upper = step = None
        if node.lower is not None:
            lower = self._eval(node.lower)
        if node.upper is not None:
            upper = self._eval(node.upper)
        if node.step is not None:
            step = self._eval(node.step)
        return slice(lower, upper, step)

    def _eval_joinedstr(self, node):
        length = 0
        evaluated_values = []
        for n in node.values:
            val = str(self._eval(n))
            if len(val) + length > MAX_STRING_LENGTH:
                raise IterableTooLong("Sorry, I will not evaluate something this long.")
            evaluated_values.append(val)
        return "".join(evaluated_values)

    def _eval_formattedvalue(self, node):
        if node.format_spec:
            fmt = "{:" + self._eval(node.format_spec) + "}"
            return fmt.format(self._eval(node.value))
        return self._eval(node.value)


class EvalWithCompoundTypes(SimpleEval):
    """
    SimpleEval with additional Compound Types, and their respective
    function editions. (list, tuple, dict, set).
    """

    _max_count = 0

    def __init__(self, operators=None, functions=None, names=None):
        super(EvalWithCompoundTypes, self).__init__(operators, functions, names)

        self.functions.update(list=list, tuple=tuple, dict=dict, set=set)

        self.nodes.update(
            {
                ast.Dict: self._eval_dict,
                ast.Tuple: self._eval_tuple,
                ast.List: self._eval_list,
                ast.Set: self._eval_set,
                ast.ListComp: self._eval_comprehension,
                ast.GeneratorExp: self._eval_comprehension,
            }
        )

    def eval(self, expr, previously_parsed=None):
        # reset _max_count for each eval run
        self._max_count = 0
        return super(EvalWithCompoundTypes, self).eval(expr, previously_parsed)

    def _eval_dict(self, node):
        return {self._eval(k): self._eval(v) for (k, v) in zip(node.keys, node.values)}

    def _eval_tuple(self, node):
        return tuple(self._eval(x) for x in node.elts)

    def _eval_list(self, node):
        return list(self._eval(x) for x in node.elts)

    def _eval_set(self, node):
        return set(self._eval(x) for x in node.elts)

    def _eval_comprehension(self, node):
        to_return = []

        extra_names = {}

        previous_name_evaller = self.nodes[ast.Name]

        def eval_names_extra(node):
            """
            Here we hide our extra scope for within this comprehension
            """
            if node.id in extra_names:
                return extra_names[node.id]
            return previous_name_evaller(node)

        self.nodes.update({ast.Name: eval_names_extra})

        def recurse_targets(target, value):
            """
                Recursively (enter, (into, (nested, name), unpacking)) = \
                             and, (assign, (values, to), each
            """
            if isinstance(target, ast.Name):
                extra_names[target.id] = value
            else:
                for t, v in zip(target.elts, value):
                    recurse_targets(t, v)

        def do_generator(gi=0):
            g = node.generators[gi]
            for i in self._eval(g.iter):
                self._max_count += 1

                if self._max_count > MAX_COMPREHENSION_LENGTH:
                    raise IterableTooLong("Comprehension generates too many elements")
                recurse_targets(g.target, i)
                if all(self._eval(iff) for iff in g.ifs):
                    if len(node.generators) > gi + 1:
                        do_generator(gi + 1)
                    else:
                        to_return.append(self._eval(node.elt))

        try:
            do_generator()
        finally:
            self.nodes.update({ast.Name: previous_name_evaller})

        return to_return


def simple_eval(expr, operators=None, functions=None, names=None):
    """Simply evaluate an expresssion"""
    s = SimpleEval(operators=operators, functions=functions, names=names)
    return s.eval(expr)
